auth/service.go

```diff
@@ -0,0 +1,233 @@
+	commonErrors "github.com/fabric8-services/fabric8-common/errors"
+	"github.com/fabric8-services/fabric8-common/log"
+	authclient "github.com/fabric8-services/fabric8-tenant/auth/client"
+	"github.com/fabric8-services/fabric8-tenant/configuration"
+	goaclient "github.com/goadesign/goa/client"
+	goajwt "github.com/goadesign/goa/middleware/security/jwt"
+	"github.com/pkg/errors"
+)
+
+type Service struct {
+	Config        *configuration.Data
+	ClientOptions []commonConfig.HTTPClientOption
+	SaToken       string
+}
+
+// NewAuthService retrieves SA OAuth token and creates a service instance that is the main point for communication with auth service
+func NewAuthService(config *configuration.Data, options ...commonConfig.HTTPClientOption) (*Service, error) {
+	c := &Service{
+		Config:        config,
+		ClientOptions: options,
+	}
+	saToken, err := c.getOAuthToken(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	c.SaToken = *saToken
+	return c, nil
+}
+
+// User contains user data retrieved from auth service and OS username and user token
+type User struct {
+	UserData           *authclient.UserDataAttributes
+	OpenshiftUsername  string
+	OpenshiftUserToken string
+}
+
+// NewUser retrieves user data from auth service related to JWT token stored in the given context.
+// It also retrieves OS username and user token for the user's cluster.
+func (s *Service) NewUser(ctx context.Context) (*User, error) {
+	userToken := goajwt.ContextJWT(ctx)
+	if userToken == nil {
+		return nil, commonErrors.NewUnauthorizedError("Missing JWT token")
+	}
+
+	// fetch the cluster the user belongs to
+	userData, err := s.GetAuthUserData(ctx, userToken)
+	if err != nil {
+		return nil, err
+	}
+
+	if userData.Cluster == nil {
+		log.Error(ctx, nil, "no cluster defined for tenant")
+		return nil, commonErrors.NewInternalError(ctx, fmt.Errorf("unable to provision to undefined cluster"))
+	}
+
+	// fetch the users cluster token
+	openshiftUsername, openshiftUserToken, err := s.ResolveUserToken(ctx, *userData.Cluster, userToken.Raw)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err":         err,
+			"cluster_url": *userData.Cluster,
+		}, "unable to fetch tenant token from auth")
+		return nil, commonErrors.NewUnauthorizedError("Could not resolve user token")
+	}
+
+	return &User{
+		UserData:           userData,
+		OpenshiftUsername:  openshiftUsername,
+		OpenshiftUserToken: openshiftUserToken,
+	}, nil
+}
+
+// GetAuthURL returns URL of auth service
+func (s *Service) GetAuthURL() string {
+	return s.Config.GetAuthURL()
+}
+
+// NewSaClient creates an instance of auth client with SA token
+func (s *Service) NewSaClient() (*authclient.Client, error) {
+	return s.newClient(s.SaToken)
+}
+
+func (s *Service) newClient(token string) (*authclient.Client, error) {
+	client, err := newClient(s.Config.GetAuthURL(), token, s.ClientOptions...)
+	if err != nil {
+		return nil, err
+	}
+	if token != "" {
+		client.SetJWTSigner(
```

Comment:
If we use Auth client only when we start the service to load keys and SA token (we will need to change it later to support dynamic keys and SA token loading) then it's fine to use JWTSigner only.
But if we also use Auth during handing some tenant requests then we also would need to forward request ID header. But we would need the original context for that.